rclpy/actions/minimal_action_client/examples_rclpy_minimal_action_client/client.py

```diff
@@ -70,11 +73,18 @@ def send_goal(self):
 def main(args=None):
     rclpy.init(args=args)
 
-    action_client = MinimalActionClient()
+    try:
+        action_client = MinimalActionClient()
 
-    action_client.send_goal()
+        action_client.send_goal()
 
-    rclpy.spin(action_client)
+        rclpy.spin(action_client)
+    except KeyboardInterrupt:
+        pass
+    except ExternalShutdownException:
+        sys.exit(1)
+    finally:
+        rclpy.try_shutdown()
```

Comment:
I know this is just following up on more changes due to https://github.com/ros2/rclpy/pull/830, but it seems like a real shame to me that it's not just:

```suggestion
        rclpy.spin(action_client)
    except KeyboardInterrupt:
        pass
```

It's a discussion for another time (maybe never given the list of things we have to do), but I don't see why the executor is throwing when shutdown (externally or otherwise):

https://github.com/ros2/rclpy/blame/99f1ce933ddd1677fda8fa602af3684b93f814eb/rclpy/rclpy/executors.py#L689C5-L689C5

In rclcpp the executor just stops and spin returns without exception. Seems like a reasonable behavior for python too, but maybe I'm missing some context.

Reply:
Also, I'm not sure why SIGTERM (or actually any shutdown of the context) results in `sys.exit(1)`.

Reply:
I talked to @sloretz about it off-line, and he doesn't remember a strong reason why shutting down the executor or the context should cause spin to throw. In my opinion it's an unnecessary deviation from how rclcpp works. I don't know if @ivanpauno is watching, but maybe he has some insight as well.

Either way, @sloretz and I both agreed that the `finally: try_shutdown()` stuff can just go, again unless someone else has a good reason for it.

Reply:
I agree with you completely; the fact that we have to do this whole dance seems incorrect to me.  Given the way that Python handles `KeyboardInterrupt`, I think we *have* to do that bit, but the rest of it seems to me like we aren't doing signal handling properly.  This also manifests in the call to `rclpy.try_shutdown()`; in my opinion, that shouldn't exist at all and `rclpy.shutdown()` should just do whatever the correct thing is (even if that means deferring shutdown because something is in use).

All of that said, I think for now we should put this in to make these examples a little quieter, and then address the signal handling separately.

Reply:
> Either way, @sloretz and I both agreed that the `finally: try_shutdown()` stuff can just go, again unless someone else has a good reason for it.

Hm.  All of our rclcpp examples use the pattern:

```
rclcpp::init();
do_stuff();
rclcpp::shutdown();
```

In rclpy, the equivalent would be:

```
rclpy.init()
try:
  do_stuff()
finally:
  rclpy.try_shutdown()
```

I don't really feel that strongly about it, but it does kind of feel like we should do the shutdown in rclpy as well, no?

Reply:
All right.  After further discussion, we agreed to remove the `try_shutdown` everywhere to make the examples simpler.  We should also remove `ExternalShutdownException`, but that will be done at a later time; see #1279.

We also discussed that this would all be a lot nicer if the various ROS 2 entities in rclpy were context managers, so I've also opened up #1280 to talk about that.

With that, I've now done all of the updates for this PR, so I could use new reviews on this.